plugin/dnstap/msg/msg.go

```diff
@@ -1,159 +1,78 @@
 package msg
 
 import (
-	"errors"
+	"fmt"
 	"net"
-	"strconv"
 	"time"
 
 	tap "github.com/dnstap/golang-dnstap"
-	"github.com/miekg/dns"
 )
 
-// Builder helps to build a Dnstap message.
-type Builder struct {
-	Packed      []byte
-	SocketProto tap.SocketProtocol
-	SocketFam   tap.SocketFamily
-	Address     net.IP
-	Port        uint32
-	TimeSec     uint64
-	TimeNsec    uint32
-
-	err error
-}
-
-// New returns a new Builder
-func New() *Builder {
-	return &Builder{}
-}
-
-// Addr adds the remote address to the message.
-func (b *Builder) Addr(remote net.Addr) *Builder {
-	if b.err != nil {
-		return b
-	}
-
-	switch addr := remote.(type) {
+// SetQueryAddress adds the query address to the message. This also sets the SocketFamily and SocketProtocol.
+func SetQueryAddress(t *tap.Message, addr net.Addr) error {
+	inet := tap.SocketFamily_INET
+	t.SocketFamily = &inet
+	switch a := addr.(type) {
 	case *net.TCPAddr:
-		b.Address = addr.IP
-		b.Port = uint32(addr.Port)
-		b.SocketProto = tap.SocketProtocol_TCP
+		t.QueryAddress = a.IP
```

Comment:
consider code de-duplication and optimization like below
```go
// these could be defined once at a packet level and shared between all messages since they are read only
var (
  protoUDP = tap.SocketProtocol_UDP
  protoTCP = tap.SocketProtocol_TCP
  familyINET = tap.SocketFamily_INET
  familyINET6 = tap.SocketFamily_INET6
)

func SetQueryAddress(t *tap.Message, addr net.Addr) error {
  var (
    ip net.IP
    port uint32
  )

  t.SocketFamily = &familyINET

  switch a := addr.(type) {
  case *net.TCPAddr:
    ip, port = a.IP, uint32(a.Port)
    t.SocketProtocol = &protoTCP
  case *net.UDPAddrr:
    ip, port = a.IP, uint32(a.Port)
    t.SocketProtocol = &protoUDP
  default:
    return fmt.Errorf("unknown address type: %T", a)
  }

  t.QueryAddress = ip
  t.QueryPort = &port
  if ip.To4() == nil {
    t.SocketFamily = &familyINET6
  }
}
```

Reply:
done. I don't have `a` outside of the type switch, so extracting out the `a.IP.To4()` becomes ikcy.